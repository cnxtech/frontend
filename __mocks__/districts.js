import { GET_DISTRICTS } from 'graphql/listings/queries'

const districts = [
  {
    request: {
      query: GET_DISTRICTS,
      variables: {},
    },
    result: {
      data: {
        districts: [
          {"citySlug":"rio-de-janeiro","stateSlug":"rj","nameSlug":"humaita","city":"Rio de Janeiro","state":"RJ","name":"Humait√°"},
          {"citySlug":"rio-de-janeiro","stateSlug":"rj","nameSlug":"copacabana","city":"Rio de Janeiro","state":"RJ","name":"Copacabana"},
          {"citySlug":"rio-de-janeiro","stateSlug":"rj","nameSlug":"botafogo","city":"Rio de Janeiro","state":"RJ","name":"Botafogo"},
          {"citySlug":"rio-de-janeiro","stateSlug":"rj","nameSlug":"catete","city":"Rio de Janeiro","state":"RJ","name":"Catete"},
          {"citySlug":"rio-de-janeiro","stateSlug":"rj","nameSlug":"cosme-velho","city":"Rio de Janeiro","state":"RJ","name":"Cosme Velho"},
          {"citySlug":"rio-de-janeiro","stateSlug":"rj","nameSlug":"flamengo","city":"Rio de Janeiro","state":"RJ","name":"Flamengo"}
        ],
      },
    },
  },
]

export {
  districts
}
