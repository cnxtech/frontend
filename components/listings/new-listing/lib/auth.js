import { getUserInfo, getPhoneParts } from 'lib/user'

const getUser = async (id, updatePhone) => {
  const userInfo = await getUserInfo(id)
  if (userInfo && !userInfo.error) {
    // Update user info in redux
    const fullPhone = getPhoneParts(userInfo.phone)
    const name = userInfo.name
    updatePhone({
      name: name,
      internationalCode: fullPhone.internationalCode,
      localAreaCode: fullPhone.localAreaCode,
      number: fullPhone.number
    })
  }

  return userInfo
}

const hasPhoneNumber = (phone) => {
  if (!phone) {
    return false
  }
  const { internationalCode, localAreaCode, number } = phone
  return (
    internationalCode !== null &&
    localAreaCode !== null &&
    number !== null
  )
}

export {
  getUser,
  hasPhoneNumber
}
