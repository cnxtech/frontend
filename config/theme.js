import theme from '@emcasa/ui'
import {fas} from '@fortawesome/pro-solid-svg-icons'
import {fal} from '@fortawesome/pro-light-svg-icons'
import icons from './icons'

export default {
  ...theme,
  icons: {emcasa: icons, default: fas, light: fal}
}
