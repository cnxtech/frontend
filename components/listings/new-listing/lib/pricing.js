import {captureException} from '@sentry/browser'
import {getAddressInput} from 'lib/address'
import {estimatePrice, getPricingInput} from 'lib/listings/pricing'
import {
  log,
  getSellerEventPrefix,
  SELLER_ONBOARDING_PRICING_SUCCESS,
  SELLER_ONBOARDING_PRICING_FAILED
} from 'lib/logging'

export const requestPricing = async (apolloClient, userInfo, {homeDetails, rooms, location, pricing, updatePricing, evaluation}) => {
  // Prepare input
  const addressInput = getAddressInput(location.addressData)
  const pricingInput = getPricingInput(addressInput, homeDetails, rooms, userInfo)

  // Run mutation
  const response = await estimatePrice(apolloClient, pricingInput)
  if (response.error) {
    captureException(new Error(response.error))
    return response
  }

  // Handle result
  if (response.result) {
    const {id, suggestedPrice, userPrice} = response.result
    if (suggestedPrice) {
      log(`${getSellerEventPrefix(evaluation)}${SELLER_ONBOARDING_PRICING_SUCCESS}`, {
        name: userInfo.name,
        phone: userInfo.phone,
        priceRequestId: id,
        pricingInput,
        suggestedPrice: suggestedPrice
      })
    } else {
      log(`${getSellerEventPrefix(evaluation)}${SELLER_ONBOARDING_PRICING_FAILED}`, {
        name: userInfo.name,
        phone: userInfo.phone,
        pricingInput
      })
    }
    updatePricing({
      ...pricing,
      priceRequestId: id,
      suggestedPrice,
      userPrice
    })
  } else {
    return {
      error: 'No response.'
    }
  }
}