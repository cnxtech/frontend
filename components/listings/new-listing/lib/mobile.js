import theme from '@emcasa/ui'

const MOBILE_BREAKPOINT_WIDTH = parseInt(theme.breakpoints[0])

const isMobile = () => {
  if (!process.browser) {
    return false
  }
  return screen.width <= MOBILE_BREAKPOINT_WIDTH
}

export {
  isMobile
}
