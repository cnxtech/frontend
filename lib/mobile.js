import theme from '@emcasa/ui'

/**
 * @deprecated. This method ends up causing a difference between what is rendered on the server and what is rendered on
 * the client. We should replace this with <Mobile> and <Desktop> components.
 */
const isMobile = (width) => {
  if (!process.browser) {
    return false
  }
  return (width || window.innerWidth) <= parseInt(theme.breakpoints[1])
}

export {
  isMobile
}
