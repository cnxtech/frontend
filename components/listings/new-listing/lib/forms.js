import {isMobile} from 'lib/mobile'

const autoFocus = (ref) => {
  if (ref && !isMobile()) {
    ref.focus()
  }
}

export {
  autoFocus
}
