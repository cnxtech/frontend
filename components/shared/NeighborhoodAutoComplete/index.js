import React, { Component } from 'react'
import Link from 'next/link'
import slug from 'slug'
import PropTypes from 'prop-types'
import {Query} from 'react-apollo'
import { MoonLoader } from 'react-spinners'
import Fuse from 'fuse.js'
import {GET_DISTRICTS} from 'graphql/listings/queries'
import {isMobile} from 'lib/mobile'
import theme from '@emcasa/ui'
import Input from '@emcasa/ui-dom/components/Input'
import Col from '@emcasa/ui-dom/components/Col'
import View from '@emcasa/ui-dom/components/View'
import Text from '@emcasa/ui-dom/components/Text'
import { normalizeLocation } from './locationNormalizer'
import {
  log,
  LISTING_SEARCH_FILTER_LOCATION
} from 'lib/amplitude'
import {
  SearchResultContainer,
  SearchResultItem,
  InputContainer,
  BackIcon
} from './styles'

export default class NeighborhoodAutoComplete extends Component {
  constructor(props) {
    super(props)
    this.timer = null
    this.searchInput = React.createRef()
    this.inputContainer = React.createRef()
    this.predictionsIds = []
    this.secondaryText = undefined
    this.onBlur = this.onBlur.bind(this)
    this.hidePredictions = this.hidePredictions.bind(this)
    this.logLocationSelection = this.logLocationSelection.bind(this)
    this.hidePredictionsTimer = null
  }

  state = {
    dirty: false,
    predictions: [],
    predictionSelected: 0,
    showPredictions: false,
    place: {},
    search: '',
    errors: [],
    loadingPlaceInfo: false
  }

  static propTypes = {
    searchResultsMargin: PropTypes.number,
    defaultValue: PropTypes.string,
    onSelectAddress: PropTypes.func.isRequired,
    onClearInput: PropTypes.func,
    onBackPressed: PropTypes.func,
    mountWithVisiblePredictions: PropTypes.bool
  }

  static defaultProps = {
    defaultValue: null
  }

  componentWillMount() {
    if (this.props.mountWithVisiblePredictions) {
      this.setState({showPredictions: true})
    }
  }

  componentDidMount() {
    if (this.searchInput.current && isMobile()) {
      this.searchInput.current.focus()
    }
  }

  componentWillUnmount() {
    clearTimeout(this.hidePredictionsTimer)
  }

  searchPlaces = async (input) => {
    this.setState({
      showPredictions: true,
      input: input,
      predictionSelected: 0,
    })
  }

  logLocationSelection(item) {
    const location = normalizeLocation(item)
    log(LISTING_SEARCH_FILTER_LOCATION, location)
  }

  getSearchResults = () => {
    return (
      <Query query={GET_DISTRICTS}>
        {({data: {districts}, loading}) => {
          if (loading) {
            return null
          }
          const cities = [
            {name: 'São Paulo', citySlug: 'sao-paulo', stateSlug: 'sp', state: 'SP'},
            {name: 'Rio de Janeiro', citySlug: 'rio-de-janeiro', stateSlug: 'rj', state: 'RJ'}
          ]

          const districtsSP = districts.filter(({stateSlug}) => stateSlug === 'sp')
          const districtsRJ = districts.filter(({stateSlug}) => stateSlug === 'rj')
          const items = cities.concat(districtsSP).concat(districtsRJ)
          const itemsToSearch = new Fuse(items, {threshold: 0.1 , keys: ['name', 'city', 'state']})
          const results = this.state.input ? itemsToSearch.search(this.state.input) : items
          return results.filter(d => d.name).map((item, index) => {
            let url = `/imoveis/${item.stateSlug}/${item.citySlug}`
            if (item.nameSlug) {
              url += `/${slug(item.nameSlug.toLowerCase())}`
            }
            return (
              <Link key={index} href={{
                pathname: url,
                asPath: url,
                search: (window && window.location && window.location.search) ? window.location.search : null
              }}>
                <SearchResultItem
                  onClick={() => {this.logLocationSelection(item)}}
                  height={this.props.height}
                >
                  <Text>{item.name}, {item.nameSlug ? item.city : item.state}</Text>
                </SearchResultItem>
              </Link>
            )
          })
        }}
      </Query>
    )
  }

  onKeyPress = (e) => {
    const {keyCode} = e
    const {
      predictions,
      predictionSelected,
      showPredictions,
      search
    } = this.state
    let newPredictionSelected = predictionSelected
    if (predictions.length > 0) {
      switch (keyCode) {
        case 9:
          search.indexOf('número') === -1
            ? this.setPlace(predictions[0])
            : this.setState({showPredictions: !showPredictions})
          break
        case 38:
          newPredictionSelected =
            predictionSelected > 0
              ? newPredictionSelected - 1
              : predictions.length - 1
          break
        case 40:
          newPredictionSelected =
            predictionSelected >= predictions.length - 1
              ? 0
              : newPredictionSelected + 1
          break
        case 13:
          const prediction = predictions.filter(
            (prediction) =>
              prediction.description ===
              this.predictionsIds[newPredictionSelected]
          )[0]
          this.setPlace(prediction)
          e.preventDefault()
          break
      }
    }

    this.setState({predictionSelected: newPredictionSelected})
  }

  onChange = (e) => {
    const {value} = e.target
    this.setState({
      dirty: true,
      search: value,
      place: {},
      errors: []
    })
    clearTimeout(this.timer)
    this.timer = setTimeout(this.searchPlaces.bind(null, value), 300)

    const { onClearInput } = this.props
    if (onClearInput) {
      if (!value || value === '') {
        onClearInput()
      }
    }
  }

  hidePredictions = () => {
    this.setState({showPredictions: false})
  }

  onBlur = () => {
    this.hidePredictionsTimer = setTimeout(this.hidePredictions, 500)
  }

  render() {
    const {
      dirty,
      place,
      search,
      loadingPlaceInfo,
      showPredictions,
      errors
    } = this.state
    const value = place.description || search
    const { onBackPressed, defaultValue, height } = this.props

    let suggestionsWidth = null
    if (this.inputContainer && this.inputContainer.current) {
      suggestionsWidth = this.inputContainer.current.offsetWidth
    }
    return (
      <div ref={this.inputContainer}>
        <InputContainer>
          {onBackPressed &&
            <BackIcon name="arrow-left" color="dark" onClick={onBackPressed} />
          }
          <Col width={1}>
            <Input
              onBlur={this.onBlur}
              style={{border: 0}}
              hideLabelView
              hideErrorView
              onFocus={() => this.setState({showPredictions: true})}
              onKeyDown={this.onKeyPress}
              type="text"
              name="street"
              ref={this.searchInput}
              value={dirty ? value : defaultValue}
              placeholder="Bairro ou Cidade"
              onChange={this.onChange}
              autoComplete="off"
              height={height ? height : 'tall'}
            />
          </Col>
          {loadingPlaceInfo &&
            <View mr={2}>
              <MoonLoader
                size={24}
                color={theme.colors.pink}
                style={{marginRight: 14}}
                loading={loadingPlaceInfo}
              />
            </View>
          }
        </InputContainer>
        {errors.length > 0 && <Text inline color="red" fontSize="small">{errors[0]}</Text>}
        {showPredictions &&
          <SearchResultContainer width={suggestionsWidth}>
            <Col width={[1]}>
              {this.getSearchResults()}
            </Col>
          </SearchResultContainer>
        }
      </div>
    )
  }
}
