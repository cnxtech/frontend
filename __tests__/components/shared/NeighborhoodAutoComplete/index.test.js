import React from 'react'
import NeighborhoodAutoComplete from 'components/shared/NeighborhoodAutoComplete'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import theme from '@emcasa/ui'
import { ThemeProvider } from 'styled-components'
import { MockedProvider } from 'react-apollo/test-utils'
import { isMobile } from 'lib/mobile'
import { districts } from '__mocks__/districts'

jest.mock('lib/mobile')
isMobile.mockImplementation(() => false)

describe('<NeighborhoodAutoComplete/>', () => {
  it('should render component on desktop', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <MockedProvider mocks={districts}>
          <NeighborhoodAutoComplete
            onBackPressed={() => {}}
            onSelectAddress={() => {}}
          />
        </MockedProvider>
      </ThemeProvider>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
