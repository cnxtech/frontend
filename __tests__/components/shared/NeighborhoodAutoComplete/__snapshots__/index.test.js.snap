// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NeighborhoodAutoComplete/> should render component 1`] = `
.c3 {
  -webkit-font-smoothing: antialiased;
  display: block;
  box-sizing: border-box;
  width: 100%;
  outline: none;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  padding: 0px 10px 0px 10px;
  height: 60px;
  color: #3c4858;
  font-family: Rubik;
  font-size: 18px;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
}

.c3:focus {
  border-radius: 4px;
  border: 1px solid;
  border-color: #1fb6ff;
  padding: 0px 10px 0px 10px;
  height: 60px;
}

.c3::-webkit-input-placeholder {
  color: #8492a6;
}

.c3::-moz-placeholder {
  color: #8492a6;
}

.c3:-ms-input-placeholder {
  color: #8492a6;
}

.c3::placeholder {
  color: #8492a6;
}

.c2 {
  box-sizing: border-box;
  width: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  position: absolute;
  z-index: 1;
  width: 100%;
  max-height: 220px;
  overflow-y: scroll;
  overflow-x: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  border: 1px solid #e0e6ed;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #ffffff;
}

.c1 {
  width: auto;
  height: 20px;
  margin: 4px 14px 0 14px;
}

@media only screen and (max-width:779px) {
  .c4 {
    max-height: 100%;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <svg
        height={20}
        version="1.1"
        viewBox="0 0 448 512"
        x={0}
        y={0}
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          fill="#3c4858"
          name="arrow-left"
          onClick={[Function]}
          type="default"
        />
      </svg>
    </div>
    <div
      className="c2"
      width={1}
    >
      <div
        className=""
      >
        <input
          autoComplete="off"
          className="c3"
          fontSize="default"
          height="tall"
          letterSpacing="default"
          name="street"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Bairro ou Cidade"
          style={
            Object {
              "border": 0,
            }
          }
          type="text"
          value={null}
        />
      </div>
    </div>
  </div>
  <div
    className="c4"
    width={null}
  >
    <div
      className="c2"
      width={
        Array [
          1,
        ]
      }
    />
  </div>
</div>
`;
